replica/transition.go

```diff
@@ -34,24 +70,154 @@ type Proposed struct {
 	block.Block
 }
 
-// IsTransition implements the `Transition` interface for the `Proposed` event.
+// IsTransition implements the `Transition` interface for the
+// `Proposed` event.
 func (proposed Proposed) IsTransition() {
 }
 
-// A PreVoted block has been signed and broadcast by another `Replica`.
+// A PreVoted block has been signed and broadcast by another
+// `Replica`.
 type PreVoted struct {
 	block.SignedPreVote
 }
 
-// IsTransition implements the `Transition` interface for the `PreVoted` event.
+// IsTransition implements the `Transition` interface for the
+// `PreVoted` event.
 func (preVoted PreVoted) IsTransition() {
 }
 
-// A PreCommitted polka has been signed and broadcast by another `Replica`.
+// A PreCommitted polka has been signed and broadcast by another
+// `Replica`.
 type PreCommitted struct {
 	block.SignedPreCommit
 }
 
-// IsTransition implements the `Transition` interface for the `PreCommitted` event.
+//TODO: should we have Commited as well?
+
+// IsTransition implements the `Transition` interface for the
+// `PreCommitted` event.
 func (preCommitted PreCommitted) IsTransition() {
 }
+
+func (buffer *transitionBuffer) Enqueue(transition Transition) {
+	switch transition := transition.(type) {
+	case Proposed:
+		buffer.initMapKey(transition.Height)
+		queue := buffer.buf[transition.Height]
+		if tran, ok := queue.peek(); ok {
+			switch tran.(type) {
+			case Proposed:
+			case PreVoted:
+				panic("You Enqueued a PreVoted before a Proposed")
+			case PreCommitted:
+				panic("You Enqueued a PreCommitted before a Proposed")
+			default:
+			}
+		}
+		queue.enqueue(transition)
+	case PreVoted:
+		buffer.initMapKey(transition.Height)
+		queue := buffer.buf[transition.Height]
+		if tran, ok := queue.peek(); ok {
+			switch tran.(type) {
+			case Proposed:
+				queue.reset()
+			case PreVoted:
```

Comment:
Why is there an empty case?

Reply:
To do nothing. I only need to reset the Queue iff there is a `Transition` from a previous step, while I assume I should keep multiple `PreVoted` (for example).

Reply:
The `default` case does nothing and will catch this case.

Reply:
You are absolutely correct, easy fix.